ipq40xx: add support for ASUS RT-ACRH17/RT-AC42U

roysjosh opened the pull request:
SOC:	IPQ4019
CPU:	Quad-core ARMv7 Processor [410fc075] revision 5 (ARMv7), cr=10c5387d
DRAM:	256 MB
NAND:	128 MiB Macronix MX30LF1G18AC
ETH:	Qualcomm Atheros QCA8075 Gigabit Switch (4x LAN, 1x WAN)
USB:	1x 3.0 (via Synopsys DesignWare DWC3 controller in the SoC)
WLAN1:	Qualcomm Atheros QCA4019 2.4GHz 802.11bgn 2x2:2
WLAN2:	Qualcomm Atheros QCA9984 5GHz 802.11nac 4x4:4
INPUT:	1x WPS, 1x Reset
LEDS:	Status, WIFI1, WIFI2, WAN (red & blue), 4x LAN

This board is very similar to the RT-ACRH13/RT-AC58U. It must be flashed
with an intermediary initramfs image, the jffs2 ubi volume deleted, and
then finally a sysupgrade with the final image performed.

Signed-off-by: Joshua Roys <roysjosh@gmail.com>

roysjosh commented:
@chunkeey thanks for your work on the '13. I bought a ton of both for cheap & I finally got around to getting the '17 working. Your comments would be appreciated.

ptpt52 commented:
see https://github.com/openwrt/openwrt/pull/1952

chunkeey commented:
That "RT-AC58U easy install" is still around in here:
https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=a76d39faf1d7bf8f896f018aa9b08ab3a821d15e

chunkeey commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -27,6 +27,7 @@ endef
 ALLWIFIBOARDS:= \
 	8dev_habanero-dvk \
 	aruba_ap-303 \
+	asus_rt-acrh17 \
```
Did you sent the boardfiles for both wifis upstream?

roysjosh replied:
TODO, I'll leave this comment open

chunkeey replied:
Ok, from what I can see, everything else looks ok. I'll wait for the posts to show up on the ML. Note: It's enough to post them. What they do with it is upstreams own choice.

roysjosh replied:
https://lists.infradead.org/pipermail/ath10k/2022-January/013217.html and https://lists.infradead.org/pipermail/ath10k/2022-January/013218.html

chunkeey commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -26,6 +26,10 @@ case "$FIRMWARE" in
 			/lib/firmware/ath10k/QCA9888/hw2.0/board.bin
 		;;
+	asus,rt-acrh17)
+		CI_UBIPART=UBI_DEV
```
For the AC58U I wanted to get rid of the UBI_DEV partition at some point. In theory if you rename the UBI_DEV to ubi and remove all the CI_UBIPART and extra bootargs it should still work as intended.

roysjosh replied:
Tested on both AC42U and AC58U.

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-rt-acrh17.dts:
```diff
@@ -0,0 +1,263 @@
+		compatible = "gpio-leds";
+
+		led_power: status {
+			label = "blue:status";
```
since 5.4-ish a new way of describing LEDs have been introduced.

https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/common.yaml

can you please add those to the led-nodes and remove the (deprecated) label, for all except this status/power node.
(Reason being here that the diag.sh get_led_dt() can't find label-less led-nodes yet.)

(Also upstream shifted to rename nodes like led-0 { ... led-1 { ... led-2 {... instead of status, wan, wlan, etc. I wonder
when they also start to rename the reset and wps buttons...)

roysjosh replied:
I didn't see a way to use the dts to set lan1-4 triggers so those are still done via startup script. Comments welcome here.

chunkeey replied:
Yes, LED triggers need the userspace scripts. Things could/will be different with DSA. But for now, this is fine.

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-rt-acrh17.dts:
```diff
@@ -0,0 +1,263 @@
+
+			partition@0 {
+				label = "Bootloader";
+				reg = <0x000000000000 0x000000400000>;
```
Hm, device-tree's addressable reg properties works on 32-bit values. Adding extra 0x0000 do not hurt but they sort of look odd when #address-cells and size-cells are both set to <1>; This <1> means that there are just one 32-bit integer for each the offset/address and size.

(Could you please chop up this big bootloader partition and enumerate all those 5-8 qcom partitions instead?)

roysjosh replied:
Done and done, found the partition layout in the vendor fw GPL source @ ./release/src-qca-dakota/uboot-1.0/tools/RT-AC82U/partition_ref/nand-partition.xml

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-rt-acrh17.dts:
```diff
@@ -0,0 +1,263 @@
+	status = "okay";
+};
+
+&vqmmc {
```
This is the MMC regulator. is there a MMC? The commit log doesn't mention it and there are no sdhci nodes.
Is this regulator used for something else?

roysjosh replied:
I saw the memory range in the stock FW /proc/iomem. Probably not in use & I'll remove.

chunkeey commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -217,6 +221,27 @@ define Device/asus_rt-ac58u
+	UIMAGE_NAME:=$(shell echo -e '\03\01\01\01RT-AC82U')
+	UIMAGE_MAGIC := 0x27051956
+	IMAGES += factory.trx
+	IMAGE/factory.trx := copy-file $(KDIR)/tmp/$$(KERNEL_INITRAMFS_IMAGE) | uImage none
```
This was tried before, but the Initramfs images is not available in the ImageBuilder. This sadly breaks services like OpenWrt's own firmware-selector https://firmware-selector.openwrt.org/ and third-party https://chef.libremesh.org/

mans0n replied:
> This was tried before, but the Initramfs images is not available in the ImageBuilder. This sadly breaks services like OpenWrt's own firmware-selector https://firmware-selector.openwrt.org/ and third-party https://chef.libremesh.org/

Too bad. :(
The same can be accomplished by `KERNEL_INITRAMFS` and `KERNEL_INITRAMFS_SUFFIX` I guess?

chunkeey replied:
The Initramfs is useful for the Bootloader's SDRAM boot option. I would hate to see that being rebranded as a "factory.trx" when it's not doing a proper installation like a factory image would. (sysupgrade now also complains about when you try to sysupgrade from a initramfs).

chunkeey commented:
It depends whenever the "easy install" will make a come-back or not.

If it does then: "yes, you could just upgrade through the web-ui". Still some instructions on how to open up the device and attach a UART/serial cable (location of the header/does it need to be soldered, pinout for GND,TX,RX,(VCC - with warning not to connect because 5V vs 3.3V) would be welcome. These are always great to have for the debricking purposes (and wiki!).

if not: please a step-by-step guide to flash it through the bootloader.

mans0n commented:
I would use RT-AC42U for its primary model name, it is much more consistent with other ASUS devices.

ptpt52 commented:
> It depends whenever the "easy install" will make a come-back or not.
> 
> If it does then: "yes, you could just upgrade through the web-ui". Still some instructions on how to open up the device and attach a UART/serial cable (location of the header/does it need to be soldered, pinout for GND,TX,RX,(VCC - with warning not to connect because 5V vs 3.3V) would be welcome. These are always great to have for the debricking purposes (and wiki!).
> 
> if not: please a step-by-step guide to flash it through the bootloader.


why we have to use an custom-uboot to flash via web ui?

chunkeey commented:
> why we have to use an custom-uboot to flash via web ui?

Oh, this uboot is placed into the kernel partition and get's overwritten in the process. It's not a replacement for the stock uboot, it leaves it untouched. I also have a script generator that creates one huge .scr. Thing is: scripts have to be started with the "source" command. And a sysupgrade.app for uboot.

roysjosh commented:
I have a few questions that I was wondering if anyone has taken a look at:

1. it looks like AHB is supposed to support MSI but [the bringup commit](https://github.com/torvalds/linux/commit/1c44fcb9234c80bbf9e4043deab8492e1c0dcf99) said there was a problem. Does anyone know if the ipq40xx AHB MSI hardware is functional?
2. it looks like wifi throughput is frequently CPU bound by ksoftirqd/0. It looks like ath10k on ipq40xx only has one interrupt (MSI for PCIe devices, legacy for AHB devices). I can see where wifi0 and wifi1 over AHB have 16x MSI interrupts & 1x legacy interrupt but I'm not sure if I should add something similar to pcie0/wifi2. Thoughts?
3. even with irqbalance/sysfs suggesting that ath10k (PCIe, specifically) interrupts be spread over all 4 cores, that doesn't happen. Is that something that can and should be fixed or is it a limitation of the platform?

```
root@RT-ACRH17:~# grep ath10k /proc/interrupts
 54:    1153604          0          0     224254     GIC-0 200 Level     ath10k_ahb
 93:    2787396          0          0          0   PCI-MSI 524288 Edge      ath10k_pci
root@RT-ACRH17:~# cat /proc/irq/93/*aff*
0    <-- affinity_hint
0    <-- effective_affinity
     <-- effective_affinity_list
f    <-- smp_affinity
0-3  <-- smp_affinity_list
```

roysjosh commented:
Well... [this thread](https://lore.kernel.org/linux-arm-kernel/20191120135852.GR25745@shell.armlinux.org.uk/T/) answered question 3. No IRQ spreading on 32-bit ARM will ever make mainline.

roysjosh commented:
For funsies I applied the "spread IRQ" patch from the linked thread.
```
root@RT-ACRH17:~# grep ath10k /proc/interrupts
 54:      12534      12239      12305      11917     GIC-0 200 Level     ath10k_ahb
 93:      38749      39828      39232      36712   PCI-MSI 524288 Edge      ath10k_pci
```

Taking an average of 5 iperf3 runs, each with two streams, gave the following:
```
    Vanilla  Spread IRQ
TX  513.2    450
RX  352      309.4
```
TX: wifi -> lan
RX: wifi <- lan

Or a loss of 12.3% in both TX and RX. :(

chunkeey commented:
> 1. it looks like AHB is supposed to support MSI but the bringup commit said there was a problem. Does anyone know if the ipq40xx AHB MSI hardware is functional?
> 2. it looks like ath10k on ipq40xx only has one interrupt (MSI for PCIe devices, legacy for AHB devices). I can see where wifi0 and wifi1 over AHB have 16x MSI interrupts & 1x legacy interrupt but I'm not sure if I should add something similar to pcie0/wifi2. Thoughts?

these are deep rabbit holes... and at the end it might not make the big difference you are after. I found that the RT-AC58U OpenWrt's wireless throughput was mostly depended on having the correct board file... which caused the ipq-wifi package
mess we are in. 

What could maybe help: On the mailing lists there is currently a discussion about making ath10k use threaded napi. 
https://lkml.org/lkml/2021/12/14/1105

roysjosh commented:
Deep rabbit holes are my favorite :D Yeah, I'm just frustrated that it seems like these devices should be a little more performant but something is holding them back. I actually applied the [patch](https://lkml.org/lkml/2021/12/14/1105) from that thread last night. I haven't seen a difference yet though I'm not 100% sure it is working.

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-rt-ac42u.dts:
```diff
@@ -0,0 +1,302 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qcom-ipq4019-ap.dk04.1.dtsi"
```
This was from ASUS's device dts right?

This would be Ok, if the ASUS ACRH17/AC42U was really based on the Qualcomm reference board. Please include ``` #include "qcom-ipq4019.dtsi" ``` instead. 

Note: (This should also get rid of SPI and the secondary uart which you probably are not using)

roysjosh replied:
Done

roysjosh committed:
ipq40xx: add support for ASUS RT-ACRH17/RT-AC42U

SOC:	IPQ4019
CPU:	Quad-core ARMv7 Processor [410fc075] revision 5 (ARMv7), cr=10c5387d
DRAM:	256 MB
NAND:	128 MiB Macronix MX30LF1G18AC
ETH:	Qualcomm Atheros QCA8075 Gigabit Switch (4x LAN, 1x WAN)
USB:	1x 3.0 (via Synopsys DesignWare DWC3 controller in the SoC)
WLAN1:	Qualcomm Atheros QCA4019 2.4GHz 802.11bgn 2x2:2
WLAN2:	Qualcomm Atheros QCA9984 5GHz 802.11nac 4x4:4
INPUT:	1x WPS, 1x Reset
LEDS:	Status, WIFI1, WIFI2, WAN (red & blue), 4x LAN

This board is very similar to the RT-ACRH13/RT-AC58U. It must be flashed
with an intermediary initramfs image, the jffs2 ubi volume deleted, and
then finally a sysupgrade with the final image performed.

Signed-off-by: Joshua Roys <roysjosh@gmail.com>

chunkeey commented:
Merged. Thanks!

roysjosh commented:
Thanks for the review!

mans0n commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -196,6 +200,32 @@ define Device/asus_map-ac2200
+	SOC := qcom-ipq4019
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	DTB_SIZE := 65536
+	IMAGE_SIZE := 20439364
```
Do we need this?
It seems `DTB_SIZE` was copied from apm821xx, but none of ipq40xx build recipe does not utilize it.
`IMAGE_SIZE` value seems weird. I would just drop it as it is not used by the image recipe anyway.

mans0n replied:
@chunkeey I found you're using the `IMAGE_SIZE` value in your latest staging tree, would you mind if I remove it?

mans0n replied:
I'll keep `IMAGE_SIZE`. I don't know if it's applicable on the artifact image (by `check-size`) but at least it doesn't hurt.
But I'll remove `DTB_SIZE` as it is not added to `DEVICE_VARS` in ipq40xx. It can be re-added later with a proper support.

mans0n commented:
@roysjosh Could you please open a new PR to fix those issues above?

roysjosh commented on target/linux/ipq40xx/base-files/etc/board.d/01_leds:
```diff
@@ -12,6 +12,12 @@ case "$board" in
 	ucidef_set_led_netdev "wan" "WAN" "amber:wan" "eth1"
 	;;
+asus,rt-ac42u)
+	ucidef_set_led_switch "lan1" "LAN1" "blue:lan" "switch0" "0x02"
+	ucidef_set_led_switch "lan2" "LAN2" "blue:lan_1" "switch0" "0x04"
+	ucidef_set_led_switch "lan3" "LAN3" "blue:lan_2" "switch0" "0x08"
+	ucidef_set_led_switch "lan4" "LAN4" "blue:lan_3" "switch0" "0x10"
```
Compiling & testing now

mans0n commented:
Never mind, I'll take care of it myself. Sorry for intruding.

I'd be happy if you can test my patch before I push it, though.
https://git.openwrt.org/?p=openwrt/staging/mans0n.git;a=summary

roysjosh commented:
@mans0n tested successfully.

Note to self: I regenerated `/etc/config/system` with a hacked-up `/bin/config_generate` that only touched the LED config & then did `/etc/init.d/led reload`.

mans0n commented:
@roysjosh Thanks for your test. In principle this would require `04_led_migration` for the led config but it's only been two days since the merge so I don't think there are many early users.

roysjosh commented:
@tiny1990 yes, similarly to the RT-ACRH13, so it requires an intermediate image + deleting a jffs2 volume + final sysupgrade. See the [`13 installation instructions](https://openwrt.org/toh/asus/rt-ac58u).

tiny1990 commented:
> 



> @tiny1990 yes, similarly to the RT-ACRH13, so it requires an intermediate image + deleting a jffs2 volume + final sysupgrade. See the [`13 installation instructions](https://openwrt.org/toh/asus/rt-ac58u).

thx for your reply, it's works for me
